blog/_posts/2016-10-11-julia-0.5-highlights.md

```diff
@@ -0,0 +1,828 @@
+For example, one might write something like:
+
+{% highlight julia %}
+max(abs(X), abs(Y))
+{% endhighlight %}
+
+This computes the absolute values of each element of `X` and `Y` and takes the larger of the corresponding elements from `abs(X)` and `abs(Y)`.
+In this form, the code allocates two temporary intermediate arrays – one to store `abs(X)` and one for `abs(Y)`.
+If we use vectorized function call syntax, however, these calls are syntactically fused into a *single* call to `broadcast` with an anonymous function.
+In other words, `max.(abs.(X), abs.(Y))` becomes
+
+{% highlight julia %}
+broadcast((x, y)->max(abs(x), abs(y)), X, Y)
+{% endhighlight %}
+
+This version of the computation avoids allocating any intermediate arrays and performs the entire vectorized computation all at once, directly into the result array.
+We can see this difference in memory usage and speed when we benchmark these expressions:
+
+{% highlight julia %}
+julia> using BenchmarkTools
+
+julia> X, Y = rand(1000,1000), rand(1000,1000);
+
+julia> @benchmark max(abs(X), abs(Y))
+BenchmarkTools.Trial:
+  memory estimate:  22.89 mb
+  minimum time:     13.95 ms (1.77% GC)
+  median time:      14.17 ms (1.76% GC)
+  mean time:        14.32 ms (1.78% GC)
+  maximum time:     17.15 ms (3.47% GC)
+
+julia> @benchmark max.(abs.(X), abs.(Y))
+BenchmarkTools.Trial:
+  memory estimate:  7.63 mb
+  minimum time:     2.84 ms (0.00% GC)
+  median time:      2.98 ms (0.00% GC)
+  mean time:        3.27 ms (18.26% GC)
+  maximum time:     5.96 ms (65.68% GC)
+
+julia> 22.89/7.63, 16.63/3.84
+(3.0,4.330729166666667)
+{% endhighlight %}
+
+I'm using the [`BenchmarkTools`](https://github.com/JuliaCI/BenchmarkTools.jl) package here instead of hand-rolled timing loops. `BenchmarkTools` has been carefully designed to avoid many of the common pitfalls of benchmarking code and to provide sound statistical estimates of how much time and memory your code uses.
+For the sake of brevity, I'm omitting some of the less relevant output from `@benchmark`.
+As you can see, allocations are reduced by a factor of 3 as predicted – only the output array is allocated, no temporaries are.
+The computation is also 4.3 times faster due to avoiding allocation and performing the entire computation in a single pass over the arrays.
+Even greater reduction in allocation can occur when we use the new `.=` operator to also do vectorized assignment:
+
+{% highlight julia %}
+julia> Z = zeros(X); # matrix of zeros similar to X
+
+julia> @benchmark Z .= max.(abs.(X), abs.(Y))
+BenchmarkTools.Trial:
+  memory estimate:  96.00 bytes
+  minimum time:     1.76 ms (0.00% GC)
+  median time:      1.82 ms (0.00% GC)
+  mean time:        1.89 ms (0.00% GC)
+  maximum time:     4.24 ms (0.00% GC)
+{% endhighlight %}
+
+With in-place vectorized assignment, we can fill the pre-allocated array, `Z`, without doing any allocation (the 96 bytes is an artifact), and do so 7.3 times faster than the original vectorized computation.
```

Comment:
That's a good point but I'm not sure explaining that is worth avoiding the artifact. People can believe that 96 bytes doesn't matter and isn't a "real" allocation.